modules/statistics/AnalyticsAdapter.js

```diff
@@ -1,130 +1,344 @@
+import {
+    TYPE_OPERATIONAL,
+    TYPE_PAGE,
+    TYPE_TRACK,
+    TYPE_UI
+} from '../../service/statistics/AnalyticsEvents';
+import { getLogger } from 'jitsi-meet-logger';
 import RTCBrowserType from '../RTC/RTCBrowserType';
 import Settings from '../settings/Settings';
 
+const MAX_CACHE_SIZE = 100;
+
+// eslist-disable-line no-undef
+const logger = getLogger(__filename);
+
 /**
- * Interface for analytics handlers.
+ * This class provides an API to lib-jitsi-meet and its users for sending
+ * analytics events. It serves as a bridge to different backend implementations
+ * ("analytics handlers") and a cache for events attempted to be sent before
+ * the analytics handlers were enabled.
+ *
+ * The API is designed to be an easy replacement for the previous version of
+ * this adapter, and is meant to be extended with more convenience methods.
+ *
+ *
+ * The API calls are translated to objects with the following structure, which
+ * are then passed to the sendEvent(event) function of the underlying handlers:
+ *
+ * {
+ *    type,
+ *
+ *    action,
+ *    actionSubject,
+ *    actionSubjectId,
+ *    attributes,
+ *    categories,
+ *    containerId,
+ *    containerType,
+ *    name,
+ *    objectId,
+ *    objectType,
+ *    source,
+ *    tags
+ * }
+ *
+ * The 'type' is one of 'operational', 'page', 'track' or 'ui', and some of the
+ * other properties are considered required according to the type.
+ *
+ * For events with type 'page', the required properties are: name.
+ *
+ * For events with type 'operational' and 'ui', the required properties are:
+ * action, actionSubject, source
+ *
+ * For events with type 'page', the required properties are:
+ * action, actionSubject, source, containerType, containerId, objectType,
+ * objectId
  */
-class AnalyticsAbstract {
+class AnalyticsAdapter {
     /**
-     *
+     * Creates new AnalyticsAdapter instance.
      */
-    sendEvent() {} // eslint-disable-line no-empty-function
-}
+    constructor() {
+        /**
+         * Whether this AnalyticsAdapter has been disposed or not.
+         * @type {boolean}
+         */
+        this.disposed = false;
+
+        /**
+         * The set of handlers to which events will be sent.
+         * @type {Set<any>}
+         */
+        this.analyticsHandlers = new Set();
```

Comment:
Could you make sure this won't completely break IE11? I have a feeling it might because it doesn't have Set.

Reply:
That's not a new addition: https://github.com/jitsi/lib-jitsi-meet/blob/master/modules/statistics/AnalyticsAdapter.js#L67